### ISSUE TITLE
The output_details and actual output of my converted TF lite model are of different shape 

### ISSUE THREAD

**Issue by DerEchteFeuerpfeil** [tensorflow/tensorflow#58844]
### 1. System information

- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): **Win10**
- TensorFlow installation (pip package or built from source): **pip**
- TensorFlow library (version, if pip package or github SHA, if built from source): **2.10.1**

### 2. Code
[Colab notebook](https://colab.research.google.com/drive/1taQXvzVbskZbk8_2IRBCgzubNDd4syhZ?usp=sharing) that includes all from model creation to testing the converted model

### 3. Problem Setting
I am trying to build a BERT-based model that does token-level prediction. So the output of my model should be (batch_size, input_length, num_classes). In my case, for a single prediction, that is (1, 128, 6).
In my understanding, higher level APIs like the [BertNLClassifier API](https://www.tensorflow.org/lite/inference_with_metadata/task_library/bert_nl_classifier) do not work for this use case as they can do just one prediction for the whole input.

That is why I have to build it myself. I used TFBert from the transformers library and put a custom classification head on top. Checking the output:
```python
print(model.output)
```
> KerasTensor(type_spec=TensorSpec(shape=(None, 128, 6), dtype=tf.float32, name=None), name='dense/BiasAdd:0', description="created by layer 'dense'")

So output is exactly what I want, great!

After checking the output shape I converted that model to TF lite format like this:
```python
converter = tf.lite.TFLiteConverter.from_keras_model(model)

converter.target_spec.supported_ops = [
  tf.lite.OpsSet.TFLITE_BUILTINS,
  tf.lite.OpsSet.SELECT_TF_OPS # required for Gelu activation function I think
]

tflite_model = converter.convert()

# Save the model.
with open('model.tflite', 'wb') as f:
  f.write(tflite_model)
```

### 4. Failure after conversion
After converting the model I ran it in my mobile Android app and was surprised when I got this error message:
> E/tf_predict: Message: Cannot copy from a TensorFlowLite tensor (StatefulPartitionedCall:0) with shape [1, 1, 8] to a Java object with shape [1, 128, 8].

So apparently my output shape changed after conversion. Strange, I thought. So after that did not work I read about the python tf lite interpreter API ```tf.lite.Interpreter```, which I then used to dig deeper for the conversion error.

So I load the model and print the output details:
```python
# Load the TFLite model and allocate tensors.
interpreter = tf.lite.Interpreter(model_path="..\dev\model_punctuation.tflite")
interpreter.allocate_tensors()

# Get input and output tensors.
input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()

print(output_details)
```
> [{'name': 'StatefulPartitionedCall:0',
  'index': 654,
  'shape': array([1, 1, 8]),
  'shape_signature': array([-1, -1,  8]),
  'dtype': numpy.float32,
  'quantization': (0.0, 0),
  'quantization_parameters': {'scales': array([], dtype=float32),
   'zero_points': array([], dtype=int32),
   'quantized_dimension': 0},
  'sparsity_parameters': {}}]

Okay I though, makes sense. The shape here is already ```array([1, 1, 8])```, so the conversion must be buggy then, right?

I went a step further and [invoked the interpreter](https://www.tensorflow.org/lite/guide/inference#load_and_run_a_model_in_python) just to check if the results match with the Android predictions:
```python
# Test the model on random input data.
input_shape = input_details[0]['shape'] # all three inputs are the same shape

input_data1 = np.array(np.zeros(input_shape), dtype=np.int32)
input_data2 = np.array(np.zeros(input_shape), dtype=np.int32)
input_data3 = np.array(np.zeros(input_shape), dtype=np.int32)

interpreter.set_tensor(input_details[0]['index'], input_data1)
interpreter.set_tensor(input_details[1]['index'], input_data2)
interpreter.set_tensor(input_details[2]['index'], input_data3)

# invoke the model
interpreter.invoke()

# The function `get_tensor()` returns a copy of the tensor data.
# Use `tensor()` in order to get a pointer to the tensor.
output_data = interpreter.get_tensor(output_details[0]['index'])
```

**Now the bug happened** that I want to describe. After invoking the interpreter, I printed the ```output_data.shape``` and it returned ```(1, 128, 8)```, which is the output shape of the original Keras model, so the "correct" output shape.

So we now have an inconsistency between the output_details and the *actual* output, which is rather confusing. I would not mind as long as I could also extract the (1, 128, 6) output from my model in my Android app, but I unfortunately couldn't.

- Why are the output shapes inconsistent?
- How can I build the TF lite model to have token-level output for my Android inference?

#### Code I used in my Android app
```kotlin
import org.tensorflow.lite.Interpreter

fun printTFTensorOutputSpecs(model:Interpreter){
	println("There are ${model.getOutputTensorCount()} many outputs")
	for (i in 0 until model.outputTensorCount){
		println("Output Tensor number $i")
		println(model.getOutputTensor(i).dataType())
		println(model.getOutputTensor(i).numDimensions())
		println(model.getOutputTensor(i).numElements())
		val iShape = model.getOutputTensor(i).shape()
		println(iShape[0].toString()+", "+iShape[1].toString()+", "+iShape[2].toString())
	}
}

// serves as input tokens, attention mask, and type_ids
val exampleInput = IntArray(128) { 0 }

val outputMap: Map<Int, Any>

try {
	tfInterpreter.allocateTensors();

	val input1 = IntBuffer.allocate(tfInterpreter.getInputTensor(0).numElements());
	val input2 = IntBuffer.allocate(tfInterpreter.getInputTensor(1).numElements());
	val input3 = IntBuffer.allocate(tfInterpreter.getInputTensor(2).numElements());

	// Populate inputs...
	input1.put(exampleInput)
	input2.put(exampleInput)
	input3.put(exampleInput)

	// check this for the other output tensors. 0 might be aggregated result
	val output = FloatBuffer.allocate(tfInterpreter.getOutputTensor(0).numElements());

	val out_1_128_8 = arrayOf(
			Array(128) { floatArrayOf(0f, 0f, 0f, 0f, 0f, 0f, 0f, 0f) }
	)

	outputMap = mapOf<Int, Any>(0 to out_1_128_8)

	printTFTensorOutputSpecs(tfInterpreter)

	// Process outputs...
	tfInterpreter.runForMultipleInputsOutputs(
			arrayOf(input1,input2,input3),
			outputMap
	)

} catch (e: Exception){
	Log.e("tf_predict", "Message: "+e.message)
}
```
----
```printTFTensorOutputSpecs(tfInterpreter)``` yielded:
> I/System.out: There are 1 many outputs
    I/System.out: Output Tensor number 0
I/System.out: FLOAT32 // dataType()
I/System.out: 3 // numDimensions()
I/System.out: 8 // numElements()
I/System.out: 1, 1, 8 // shape()

And ```tfInterpreter.runForMultipleInputsOutputs``` threw this error:
> E/tf_predict: Message: Cannot copy from a TensorFlowLite tensor (StatefulPartitionedCall:0) with shape [1, 1, 8] to a Java object with shape [1, 128, 8].

**Comment by mohantym** [comment]
Hi @DerEchteFeuerpfeil !
This is a known bug in TFLite conversion process. 
Usual work around for these  issues are to use [model binding](https://www.tensorflow.org/lite/inference_with_metadata/codegen#mlbinding)/[meta data writer](https://www.tensorflow.org/lite/models/convert/metadata_writer_tutorial)/ [signature](https://www.tensorflow.org/lite/guide/signatures) to avoid distortion of input/output shape during conversion.

Did you check in 2.11 /nightly version too ?. A toy Bert model from your side would be handy to replicate this issue.

Thank you!

**Comment by DerEchteFeuerpfeil** [comment]
Hey @mohantym,
all the necessary code is provided in the colab notebook that I linked in the very top of my issue. It includes the model, the model saving, the conversion and the testing.

TF 2.11 still fails, but I just tested it with tf-nightly `2.12.0-dev20221213` and it actually works there! (See screenshots)

![image](https://user-images.githubusercontent.com/57487741/207415327-22653e99-e5f4-4659-b4e0-f7552e0f0082.png)
![image](https://user-images.githubusercontent.com/57487741/207415377-c5a3c2d7-64a1-45e7-8f4a-fd148cc065f2.png)

Marking this as closed then?
